libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -0,0 +1,275 @@
+		{
+			auto const it = std::find_if(callArgumentNames.cbegin(), callArgumentNames.cend(), [&](ASTPointer<ASTString> const& _argName) {
+				return *_argName == parameterName;
+			});
+
+			solAssert(it != callArgumentNames.cend(), "");
+			arguments.push_back(callArguments[std::distance(callArgumentNames.begin(), it)]);
+		}
+
+	solUnimplementedAssert(!functionType->bound(), "");
+	switch (functionType->kind())
+	{
+	case FunctionType::Kind::Internal:
+	{
+		vector<string> args;
+		for (unsigned i = 0; i < arguments.size(); ++i)
+		{
+			arguments[i]->accept(*this);
+			// TODO convert
+			//utils().convertType(*arguments[i]->annotation().type, *function.parameterTypes()[i]);
+			args.emplace_back(m_context.variable(*arguments[i]));
+		}
+
+		if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+		{
+			solAssert(!functionType->bound(), "");
+			if (auto functionDef = dynamic_cast<FunctionDefinition const*>(identifier->annotation().referencedDeclaration))
+			{
+				// @TODO The function can very well return multiple vars.
+				m_code <<
+					"let " <<
+					m_context.variable(_functionCall) <<
+					" := " <<
+					m_context.virtualFunctionName(*functionDef) <<
+					"(" <<
+					joinHumanReadable(args) <<
+					")\n";
+				return false;
+			}
+		}
+
+		_functionCall.expression().accept(*this);
+
+		// @TODO The function can very well return multiple vars.
+		args = vector<string>{m_context.variable(_functionCall.expression())} + args;
+		m_code <<
+			"let " <<
+			m_context.variable(_functionCall) <<
+			" := " <<
+			m_context.internalDispatch(functionType->parameterTypes().size(), functionType->returnParameterTypes().size()) <<
+			"(" <<
+			joinHumanReadable(args) <<
+			")\n";
+		break;
+	}
+	default:
+		solUnimplemented("");
+	}
+	return false;
+}
+
+// Preliminary implementation that doesn't take care of many things like
+// name collisions etc. Mainly used for test cases.
+bool IRGeneratorForStatements::visit(InlineAssembly const& _inlineAsm)
+{
+	auto const& loc = _inlineAsm.location();
+
+	solAssert(loc.start != 1 && loc.end != -1, "Expected source location");
+
+	vector<string> params, retParams, args;
+	vector<map<string, string>> assignVars;
+	// Prepare parameters and arguments
+	for (auto const& pair: _inlineAsm.annotation().externalReferences)
+	{
+		auto varDecl = dynamic_cast<VariableDeclaration const*>(pair.second.declaration);
+		solUnimplementedAssert(!!varDecl, "");
+		solUnimplementedAssert(
+			pair.second.isOffset == false && pair.second.isSlot == false,
+			""
+		);
+
+		args.emplace_back(m_context.variableName(*varDecl));
+		params.emplace_back(varDecl->name());
+		retParams.emplace_back("ret_" + varDecl->name());
+		assignVars.emplace_back(
+			map<string, string>{
+				{"retParam", retParams.back()},
+				{"param", params.back()}
+			}
+		);
+	}
+
+	// Skip the "assembly" literal
+	int blockStart = loc.start + string("assembly").size();
```

Comment:
We should probably also assert that the source location is large enough for this.